File: apps/mobile/src/modules/photo-viewer/photoInfoModel.ts
import { getIntlLocale, translate } from '@/i18n'
import type { GalleryExif, GalleryFujiRecipe, GalleryPhoto } from '@/modules/galleries/types'

import type { Translator } from './photoInfoFormatters'
import {
  cleanRecipeValue,
  decimalCoordinate,
  formatAltitude,
  formatCoordinate,
  formatDate,
  formatEv,
  formatFilmMode,
  formatFujiDynamicRange,
  formatFujiWhiteBalance,
  formatMired,
  formatPercentage,
  textValue,
  translateExifValue,
} from './photoInfoFormatters'
import type { PhotoInfoGear } from './photoInfoGear'
import { buildPhotoInfoGear } from './photoInfoGear'

export interface PhotoInfoRow {
  id: string
  label: string
  value: string
}

export interface PhotoInfoSection {
  id: string
  title: string
  summary: string | null
  rows: PhotoInfoRow[]
}

export interface PhotoInfoMapLocation {
  latitude: number
  longitude: number
}

export interface PhotoInfoSheetModel {
  gear: PhotoInfoGear
  description: string | null
  emptyMessage: string | null
  histogramUrl: string | null
  mapLocation: PhotoInfoMapLocation | null
  place: string | null
  sections: PhotoInfoSection[]
  tags: string[]
}

type NullablePhotoInfoRow = Omit<PhotoInfoRow, 'value'> & { value: string | null }

function createRows(rows: NullablePhotoInfoRow[]): PhotoInfoRow[] {
  return rows.filter((row): row is PhotoInfoRow => row.value !== null)
}

function createSection(
  id: string,
  title: string,
  rows: PhotoInfoRow[],
  summary: string | null = null,
): PhotoInfoSection | null {
  return rows.length > 0 ? { id, title, summary, rows } : null
}

function buildExposureSection(exif: GalleryExif, t: Translator): PhotoInfoSection | null {
  // A Fuji recipe reports white balance with colour temperature, so the plain
  // EXIF row would repeat it less precisely.
  const hasRecipe = exif.FujiRecipe !== undefined

  return createSection(
    'exposure',
    t('mobile.photoInfo.exposure'),
    createRows([
      {
        id: 'exposure-program',
        label: t('exif.exposureprogram.title'),
        value: translateExifValue(t, 'exif.exposureprogram', exif.ExposureProgram),
      },
      {
        id: 'exposure-mode',
        label: t('exif.exposure.mode.title'),
        value: translateExifValue(t, 'exif.exposure.mode', exif.ExposureMode),
      },
      {
        id: 'metering-mode',
        label: t('exif.metering.mode.type'),
        value: translateExifValue(t, 'exif.metering.mode', exif.MeteringMode),
      },
      {
        id: 'white-balance',
        label: t('exif.white.balance.title'),
        value: hasRecipe ? null : translateExifValue(t, 'exif.white.balance', exif.WhiteBalance),
      },
      {
        id: 'white-balance-bias',
        label: t('exif.white.balance.bias'),
        value: formatMired(exif.WhiteBalanceBias),
      },
      { id: 'white-balance-ab', label: t('exif.white.balance.shift.ab'), value: textValue(exif.WBShiftAB) },
      { id: 'white-balance-gm', label: t('exif.white.balance.shift.gm'), value: textValue(exif.WBShiftGM) },
      { id: 'flash', label: t('exif.flash.title'), value: translateExifValue(t, 'exif.flash', exif.Flash) },
      { id: 'flash-metering', label: t('exif.flash.metering.mode'), value: textValue(exif.FlashMeteringMode) },
      {
        id: 'light-source',
        label: t('exif.light.source.type'),
        value: translateExifValue(t, 'exif.light.source', exif.LightSource),
      },
      {
        id: 'scene-type',
        label: t('exif.scene.capture.type'),
        value: translateExifValue(t, 'exif.scene.capture', exif.SceneCaptureType),
      },
      { id: 'brightness', label: t('exif.brightness.value'), value: formatEv(exif.BrightnessValue) },
      {
        id: 'sensing-method',
        label: t('exif.sensing.method.type'),
        value: translateExifValue(t, 'exif.sensing.method', exif.SensingMethod),
      },
      {
        id: 'focal-plane',
        label: t('exif.focal.plane.resolution'),
        value: (() => {
          const x = textValue(exif.FocalPlaneXResolution)
          const y = textValue(exif.FocalPlaneYResolution)
          return x || y ? `${x ?? '—'} × ${y ?? '—'}` : null
        })(),
      },
    ]),
    translateExifValue(t, 'exif.exposureprogram', exif.ExposureProgram),
  )
}

function buildFujiSection(recipe: GalleryFujiRecipe | undefined, t: Translator): PhotoInfoSection | null {
  if (!recipe) {
    return null
  }

  return createSection(
    'fuji-recipe',
    t('mobile.photoInfo.filmSimulation'),
    createRows([
      { id: 'film-mode', label: t('exif.film.mode'), value: formatFilmMode(recipe.FilmMode) },
      { id: 'dynamic-range', label: t('exif.dynamic.range'), value: formatFujiDynamicRange(recipe, t) },
      { id: 'white-balance', label: t('exif.white.balance.title'), value: formatFujiWhiteBalance(recipe, t) },
      { id: 'highlight-tone', label: t('exif.highlight.tone'), value: cleanRecipeValue(recipe.HighlightTone) },
      { id: 'shadow-tone', label: t('exif.shadow.tone'), value: cleanRecipeValue(recipe.ShadowTone) },
      { id: 'saturation', label: t('exif.saturation'), value: cleanRecipeValue(recipe.Saturation) },
      {
        id: 'sharpness',
        label: t('exif.sharpness'),
        value: translateExifValue(t, 'exif.fujirecipe-sharpness', recipe.Sharpness),
      },
      { id: 'noise-reduction', label: t('exif.noise.reduction'), value: cleanRecipeValue(recipe.NoiseReduction) },
      { id: 'clarity', label: t('exif.clarity'), value: textValue(recipe.Clarity) },
      {
        id: 'color-chrome',
        label: t('exif.color.effect'),
        value: translateExifValue(t, 'exif.fujirecipe-colorchromeeffect', recipe.ColorChromeEffect),
      },
      {
        id: 'color-chrome-blue',
        label: t('exif.blue.color.effect'),
        value: translateExifValue(t, 'exif.fujirecipe-colorchromefxblue', recipe.ColorChromeFxBlue),
      },
      {
        id: 'white-balance-fine-tune',
        label: t('exif.white.balance.fine.tune'),
        value: textValue(recipe.WhiteBalanceFineTune),
      },
      {
        id: 'grain-intensity',
        label: t('exif.grain.effect.intensity'),
        value: textValue(recipe.GrainEffectRoughness),
      },
      {
        id: 'grain-size',
        label: t('exif.grain.effect.size'),
        value: translateExifValue(t, 'exif.fujirecipe-graineffectsize', recipe.GrainEffectSize),
      },
    ]),
    formatFilmMode(recipe.FilmMode),
  )
}

function buildToneSection(photo: GalleryPhoto, t: Translator): PhotoInfoSection | null {
  const toneAnalysis = photo.toneAnalysis
  if (!toneAnalysis) {
    return null
  }

  const brightness = formatPercentage(toneAnalysis.brightness, 1)
  return createSection(
    'tone',
    t('mobile.photoInfo.tone'),
    createRows([
      { id: 'brightness', label: t('exif.brightness.title'), value: brightness },
      { id: 'contrast', label: t('exif.contrast.title'), value: formatPercentage(toneAnalysis.contrast, 1) },
      { id: 'shadow-ratio', label: t('exif.shadow.ratio'), value: formatPercentage(toneAnalysis.shadowRatio, 100) },
      {
        id: 'highlight-ratio',
        label: t('exif.highlight.ratio'),
        value: formatPercentage(toneAnalysis.highlightRatio, 100),
      },
    ]),
    brightness,
  )
}

function buildLocationSection(photo: GalleryPhoto, exif: GalleryExif | null, t: Translator): PhotoInfoSection | null {
  const altitude = formatAltitude(exif)
  return createSection(
    'location',
    t('mobile.photoInfo.locationDetail'),
    createRows([
      {
        id: 'latitude',
        label: t('exif.gps.latitude'),
        value: formatCoordinate(exif?.GPSLatitude ?? photo.location?.latitude, exif?.GPSLatitudeRef),
      },
      {
        id: 'longitude',
        label: t('exif.gps.longitude'),
        value: formatCoordinate(exif?.GPSLongitude ?? photo.location?.longitude, exif?.GPSLongitudeRef),
      },
      { id: 'altitude', label: t('exif.gps.altitude'), value: altitude },
      { id: 'address', label: t('exif.gps.address'), value: photo.location?.locationName ?? null },
    ]),
    altitude,
  )
}

function buildFileSection(
  photo: GalleryPhoto,
  exif: GalleryExif | null,
  t: Translator,
  locale: string,
): PhotoInfoSection | null {
  const filename = textValue(photo.title)
  return createSection(
    'file',
    t('mobile.photoInfo.file'),
    createRows([
      { id: 'filename', label: t('exif.filename'), value: filename },
      {
        id: 'color-space',
        label: t('exif.color.space'),
        value: translateExifValue(t, 'exif.colorspace', exif?.ColorSpace),
      },
      {
        id: 'capture-time',
        label: t('exif.capture.time'),
        value: formatDate(exif?.DateTimeOriginal ?? photo.dateTaken, locale),
      },
      { id: 'time-zone', label: t('exif.time.zone'), value: textValue(exif?.zone ?? exif?.tz) },
    ]),
    filename,
  )
}

function buildAttributionSection(exif: GalleryExif | null, t: Translator): PhotoInfoSection | null {
  const artist = textValue(exif?.Artist)
  return createSection(
    'attribution',
    t('mobile.photoInfo.attribution'),
    createRows([
      { id: 'artist', label: t('exif.artist'), value: artist },
      { id: 'copyright', label: t('exif.copyright'), value: textValue(exif?.Copyright) },
      { id: 'software', label: t('exif.software'), value: textValue(exif?.Software) },
    ]),
    artist,
  )
}

function buildMapLocation(photo: GalleryPhoto, exif: GalleryExif | null): PhotoInfoMapLocation | null {
  const latitude = decimalCoordinate(exif?.GPSLatitude ?? photo.location?.latitude, exif?.GPSLatitudeRef, 90)
  const longitude = decimalCoordinate(exif?.GPSLongitude ?? photo.location?.longitude, exif?.GPSLongitudeRef, 180)

  return latitude === null || longitude === null ? null : { latitude, longitude }
}

function buildPlace(photo: GalleryPhoto): string | null {
  const city = textValue(photo.location?.city) ?? textValue(photo.city)
  const country = textValue(photo.location?.country)
  if (city && country && city.localeCompare(country, undefined, { sensitivity: 'accent' }) === 0) {
    return city
  }
  return [city, country].filter(Boolean).join(', ') || null
}

export function buildPhotoInfoSheetModel(
  photo: GalleryPhoto,
  t: Translator = translate,
  locale = getIntlLocale(),
): PhotoInfoSheetModel {
  const exif = photo.exif ?? null
  const sections = [
    ...(exif ? [buildExposureSection(exif, t), buildFujiSection(exif.FujiRecipe, t)] : []),
    buildToneSection(photo, t),
    buildLocationSection(photo, exif, t),
    buildFileSection(photo, exif, t, locale),
    buildAttributionSection(exif, t),
  ].filter((section): section is PhotoInfoSection => section !== null)

  return {
    gear: buildPhotoInfoGear(photo, exif, t, locale),
    description: photo.description || null,
    emptyMessage: exif ? null : t('photo.noExif'),
    histogramUrl: photo.toneAnalysis ? photo.thumbnailUrl : null,
    mapLocation: buildMapLocation(photo, exif),
    place: buildPlace(photo),
    sections,
    tags: photo.tags,
  }
}

File: apps/mobile/modules/photo-masonry/ios/Models/PhotoInfoModel.swift
import Foundation

struct PhotoInfoRow: Codable, Equatable, Identifiable, Sendable {
  let id: String
  let label: String
  let value: String
}

struct PhotoInfoSection: Codable, Equatable, Identifiable, Sendable {
  let id: String
  let title: String
  let summary: String?
  let rows: [PhotoInfoRow]
}

struct PhotoInfoMapLocation: Codable, Equatable, Sendable {
  let latitude: Double
  let longitude: Double
}

struct PhotoInfoSheetModel: Codable, Equatable, Sendable {
  let gear: PhotoInfoGear
  let description: String?
  let emptyMessage: String?
  let histogramUrl: String?
  let mapLocation: PhotoInfoMapLocation?
  let place: String?
  let sections: [PhotoInfoSection]
  let tags: [String]

  static let empty = PhotoInfoSheetModel(
    gear: PhotoInfoGear(
      model: "",
      formatBadge: nil,
      styleBadge: nil,
      lens: nil,
      rating: 0,
      specs: [],
      tone: nil,
      exposure: []
    ),
    description: nil,
    emptyMessage: nil,
    histogramUrl: nil,
    mapLocation: nil,
    place: nil,
    sections: [],
    tags: []
  )
}

enum PhotoInfoModel {
  static func build(
    photo: GalleryPhoto,
    localeIdentifier: String,
    timeZone: TimeZone = .current
  ) -> PhotoInfoSheetModel {
    let exif = photo.exif
    var sections: [PhotoInfoSection] = []
    if let exif {
      if let section = exposureSection(exif: exif) {
        sections.append(section)
      }
      if let section = fujiSection(recipe: exif["FujiRecipe"]?.object) {
        sections.append(section)
      }
    }
    if let section = toneSection(photo: photo) {
      sections.append(section)
    }
    if let section = locationSection(photo: photo, exif: exif) {
      sections.append(section)
    }
    if let section = fileSection(
      photo: photo,
      exif: exif,
      localeIdentifier: localeIdentifier,
      timeZone: timeZone
    ) {
      sections.append(section)
    }
    if let section = attributionSection(exif: exif) {
      sections.append(section)
    }
    return PhotoInfoSheetModel(
      gear: PhotoInfoGear.build(
        photo: photo,
        exif: exif,
        localeIdentifier: localeIdentifier
      ),
      description: photo.description.isEmpty ? nil : photo.description,
      emptyMessage: exif == nil ? String(localized: "No embedded EXIF metadata is available for this photo.") : nil,
      histogramUrl: photo.toneAnalysis == nil ? nil : photo.thumbnailUrl,
      mapLocation: mapLocation(photo: photo, exif: exif),
      place: place(photo: photo),
      sections: sections,
      tags: photo.tags
    )
  }

  private static func exposureSection(
    exif: GalleryExif
  ) -> PhotoInfoSection? {
    let hasRecipe = exif["FujiRecipe"] != nil
    let summary = PhotoInfoFormatters.translatedExifValue(
      prefix: "exif.exposureprogram",
      value: exif["ExposureProgram"]
    )
    let x = PhotoInfoFormatters.text(exif["FocalPlaneXResolution"])
    let y = PhotoInfoFormatters.text(exif["FocalPlaneYResolution"])
    return section(
      id: "exposure",
      title: String(localized: "Exposure & Metering"),
      summary: summary,
      rows: [
        row(
          id: "exposure-program",
          label: String(localized: "Exposure Program"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.exposureprogram",
            value: exif["ExposureProgram"]
          )
        ),
        row(
          id: "exposure-mode",
          label: String(localized: "Exposure Mode"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.exposure.mode",
            value: exif["ExposureMode"]
          )
        ),
        row(
          id: "metering-mode",
          label: String(localized: "Metering Mode"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.metering.mode",
            value: exif["MeteringMode"]
          )
        ),
        row(
          id: "white-balance",
          label: String(localized: "White Balance"),
          value: hasRecipe ? nil : PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.white.balance",
            value: exif["WhiteBalance"]
          )
        ),
        row(
          id: "white-balance-bias",
          label: String(localized: "White Balance Bias"),
          value: PhotoInfoFormatters.formatMired(exif["WhiteBalanceBias"])
        ),
        row(
          id: "white-balance-ab",
          label: String(localized: "White Balance Shift (Amber-Blue)"),
          value: PhotoInfoFormatters.text(exif["WBShiftAB"])
        ),
        row(
          id: "white-balance-gm",
          label: String(localized: "White Balance Shift (Green-Magenta)"),
          value: PhotoInfoFormatters.text(exif["WBShiftGM"])
        ),
        row(
          id: "flash",
          label: String(localized: "Flash"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.flash",
            value: exif["Flash"]
          )
        ),
        row(
          id: "flash-metering",
          label: String(localized: "Flash Metering Mode"),
          value: PhotoInfoFormatters.text(exif["FlashMeteringMode"])
        ),
        row(
          id: "light-source",
          label: String(localized: "Light Source"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.light.source",
            value: exif["LightSource"]
          )
        ),
        row(
          id: "scene-type",
          label: String(localized: "Scene Capture Type"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.scene.capture",
            value: exif["SceneCaptureType"]
          )
        ),
        row(
          id: "brightness",
          label: String(localized: "Brightness Value"),
          value: PhotoInfoFormatters.formatEV(exif["BrightnessValue"])
        ),
        row(
          id: "sensing-method",
          label: String(localized: "Sensing Method"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.sensing.method",
            value: exif["SensingMethod"]
          )
        ),
        row(
          id: "focal-plane",
          label: String(localized: "Focal Plane Resolution"),
          value: x == nil && y == nil ? nil : "\(x ?? "—") × \(y ?? "—")"
        ),
      ]
    )
  }

  private static func fujiSection(
    recipe: [String: JSONValue]?
  ) -> PhotoInfoSection? {
    guard let recipe else { return nil }
    return section(
      id: "fuji-recipe",
      title: String(localized: "Film Simulation"),
      summary: PhotoInfoFormatters.formatFilmMode(recipe["FilmMode"]),
      rows: [
        row(
          id: "film-mode",
          label: String(localized: "Film Mode"),
          value: PhotoInfoFormatters.formatFilmMode(recipe["FilmMode"])
        ),
        row(
          id: "dynamic-range",
          label: String(localized: "Dynamic Range"),
          value: PhotoInfoFormatters.formatFujiDynamicRange(recipe)
        ),
        row(
          id: "white-balance",
          label: String(localized: "White Balance"),
          value: PhotoInfoFormatters.formatFujiWhiteBalance(recipe)
        ),
        row(
          id: "highlight-tone",
          label: String(localized: "Highlight Tone"),
          value: PhotoInfoFormatters.cleanRecipeValue(recipe["HighlightTone"])
        ),
        row(
          id: "shadow-tone",
          label: String(localized: "Shadow Tone"),
          value: PhotoInfoFormatters.cleanRecipeValue(recipe["ShadowTone"])
        ),
        row(
          id: "saturation",
          label: String(localized: "Saturation"),
          value: PhotoInfoFormatters.cleanRecipeValue(recipe["Saturation"])
        ),
        row(
          id: "sharpness",
          label: String(localized: "Sharpness"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.fujirecipe-sharpness",
            value: recipe["Sharpness"]
          )
        ),
        row(
          id: "noise-reduction",
          label: String(localized: "Noise Reduction"),
          value: PhotoInfoFormatters.cleanRecipeValue(recipe["NoiseReduction"])
        ),
        row(
          id: "clarity",
          label: String(localized: "Clarity"),
          value: PhotoInfoFormatters.text(recipe["Clarity"])
        ),
        row(
          id: "color-chrome",
          label: String(localized: "Color Effect"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.fujirecipe-colorchromeeffect",
            value: recipe["ColorChromeEffect"]
          )
        ),
        row(
          id: "color-chrome-blue",
          label: String(localized: "Blue Color Effect"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.fujirecipe-colorchromefxblue",
            value: recipe["ColorChromeFxBlue"]
          )
        ),
        row(
          id: "white-balance-fine-tune",
          label: String(localized: "White Balance Fine Tune"),
          value: PhotoInfoFormatters.text(recipe["WhiteBalanceFineTune"])
        ),
        row(
          id: "grain-intensity",
          label: String(localized: "Grain Effect Intensity"),
          value: PhotoInfoFormatters.text(recipe["GrainEffectRoughness"])
        ),
        row(
          id: "grain-size",
          label: String(localized: "Grain Effect Size"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.fujirecipe-graineffectsize",
            value: recipe["GrainEffectSize"]
          )
        ),
      ]
    )
  }

  private static func toneSection(
    photo: GalleryPhoto
  ) -> PhotoInfoSection? {
    guard let tone = photo.toneAnalysis else { return nil }
    let brightness = PhotoInfoFormatters.formatPercentage(tone.brightness, scale: 1)
    return section(
      id: "tone",
      title: String(localized: "Tone"),
      summary: brightness,
      rows: [
        row(id: "brightness", label: String(localized: "Brightness"), value: brightness),
        row(
          id: "contrast",
          label: String(localized: "Contrast"),
          value: PhotoInfoFormatters.formatPercentage(tone.contrast, scale: 1)
        ),
        row(
          id: "shadow-ratio",
          label: String(localized: "Shadow Ratio"),
          value: PhotoInfoFormatters.formatPercentage(tone.shadowRatio, scale: 100)
        ),
        row(
          id: "highlight-ratio",
          label: String(localized: "Highlight Ratio"),
          value: PhotoInfoFormatters.formatPercentage(tone.highlightRatio, scale: 100)
        ),
      ]
    )
  }

  private static func locationSection(
    photo: GalleryPhoto,
    exif: GalleryExif?
  ) -> PhotoInfoSection? {
    let altitude = PhotoInfoFormatters.formatAltitude(exif)
    return section(
      id: "location",
      title: String(localized: "Location Details"),
      summary: altitude,
      rows: [
        row(
          id: "latitude",
          label: String(localized: "Latitude"),
          value: PhotoInfoFormatters.formatCoordinate(
            exif?["GPSLatitude"] ?? photo.location?.latitude.map(JSONValue.number),
            reference: exif?["GPSLatitudeRef"]
          )
        ),
        row(
          id: "longitude",
          label: String(localized: "Longitude"),
          value: PhotoInfoFormatters.formatCoordinate(
            exif?["GPSLongitude"] ?? photo.location?.longitude.map(JSONValue.number),
            reference: exif?["GPSLongitudeRef"]
          )
        ),
        row(id: "altitude", label: String(localized: "Altitude"), value: altitude),
        row(
          id: "address",
          label: String(localized: "Address"),
          value: photo.location?.locationName
        ),
      ]
    )
  }

  private static func fileSection(
    photo: GalleryPhoto,
    exif: GalleryExif?,
    localeIdentifier: String,
    timeZone: TimeZone
  ) -> PhotoInfoSection? {
    let filename = PhotoInfoFormatters.text(photo.title)
    return section(
      id: "file",
      title: String(localized: "Image & File"),
      summary: filename,
      rows: [
        row(id: "filename", label: String(localized: "Filename"), value: filename),
        row(
          id: "color-space",
          label: String(localized: "Color Space"),
          value: PhotoInfoFormatters.translatedExifValue(
            prefix: "exif.colorspace",
            value: exif?["ColorSpace"]
          )
        ),
        row(
          id: "capture-time",
          label: String(localized: "Capture Time"),
          value: PhotoInfoFormatters.formatDate(
            PhotoInfoFormatters.text(exif?["DateTimeOriginal"]) ?? photo.dateTaken,
            localeIdentifier: localeIdentifier,
            timeZone: timeZone
          )
        ),
        row(
          id: "time-zone",
          label: String(localized: "Time Zone"),
          value: PhotoInfoFormatters.text(exif?["zone"] ?? exif?["tz"])
        ),
      ]
    )
  }

  private static func attributionSection(
    exif: GalleryExif?
  ) -> PhotoInfoSection? {
    let artist = PhotoInfoFormatters.text(exif?["Artist"])
    return section(
      id: "attribution",
      title: String(localized: "Attribution"),
      summary: artist,
      rows: [
        row(id: "artist", label: String(localized: "Artist"), value: artist),
        row(
          id: "copyright",
          label: String(localized: "Copyright"),
          value: PhotoInfoFormatters.text(exif?["Copyright"])
        ),
        row(
          id: "software",
          label: String(localized: "Software"),
          value: PhotoInfoFormatters.text(exif?["Software"])
        ),
      ]
    )
  }

  private static func mapLocation(photo: GalleryPhoto, exif: GalleryExif?) -> PhotoInfoMapLocation? {
    let latitude = PhotoInfoFormatters.decimalCoordinate(
      exif?["GPSLatitude"] ?? photo.location?.latitude.map(JSONValue.number),
      reference: exif?["GPSLatitudeRef"],
      limit: 90
    )
    let longitude = PhotoInfoFormatters.decimalCoordinate(
      exif?["GPSLongitude"] ?? photo.location?.longitude.map(JSONValue.number),
      reference: exif?["GPSLongitudeRef"],
      limit: 180
    )
    guard let latitude, let longitude else { return nil }
    return PhotoInfoMapLocation(latitude: latitude, longitude: longitude)
  }

  private static func place(photo: GalleryPhoto) -> String? {
    let city = PhotoInfoFormatters.text(photo.location?.city) ?? PhotoInfoFormatters.text(photo.city)
    let country = PhotoInfoFormatters.text(photo.location?.country)
    if let city, let country, city.compare(country, options: .caseInsensitive) == .orderedSame {
      return city
    }
    return [city, country].compactMap { $0 }.joined(separator: ", ").trimmingToNil
  }

  private static func row(id: String, label: String, value: String?) -> PhotoInfoRow? {
    value.map { PhotoInfoRow(id: id, label: label, value: $0) }
  }

  private static func section(
    id: String,
    title: String,
    summary: String?,
    rows: [PhotoInfoRow?]
  ) -> PhotoInfoSection? {
    let rows = rows.compactMap { $0 }
    return rows.isEmpty ? nil : PhotoInfoSection(id: id, title: title, summary: summary, rows: rows)
  }
}
